### PR TITLE
Fixed when:evaluate on non-standard (non-CI*) env vars

### PR DESCRIPTION
**Pull request by zc-devs**
Makes it possible to evaluate `when` constraint on custom environment variables.

Workaround for #1668.

### CODE REVIEW

**Comment by zc-devs** [comment]
Trace of environment variables propagation.

server/router/api.go#L82
repo.POST("/pipelines", session.MustPush, api.CreatePipeline)

server/api/pipeline.go#L50
func CreatePipeline(c *gin.Context) {

server/api/pipeline.go#L66
tmpBuild := createTmpPipeline(model.EventManual, lastCommit, repo, user, &opts)

server/api/pipeline.go#L76
func createTmpPipeline(event model.WebhookEvent, commitSHA string, repo *model.Repo, user *model.User, opts *model.PipelineOptions) *model.Pipeline {

server/api/pipeline.go#L87
AdditionalVariables: opts.Variables,
The env vars from frontend when we run pipeline manually

server/api/pipeline.go#L68
pl, err := pipeline.Create(c, _store, repo, tmpBuild)

server/pipeline/create.go#L32
func Create(ctx context.Context, _store store.Store, repo *model.Repo, pipeline *model.Pipeline) (*model.Pipeline, error) {

server/pipeline/create.go#L66
filtered, parseErr = checkIfFiltered(repo, pipeline, forgeYamlConfigs)

server/pipeline/filter.go#L56
func checkIfFiltered(repo *model.Repo, p *model.Pipeline, forgeYamlConfigs []*forge_types.FileMeta) (bool, error) {

server/pipeline/filter.go#L75
if match, err := parsedPipelineConfig.When.Match(matchMetadata, true, p.AdditionalVariables); !match && err == nil {
Ignore if the pipeline was filtered by matched constraints

pipeline/frontend/yaml/constraint/constraint.go#L66
func (when *When) Match(metadata metadata.Metadata, global bool, env map[string]string) (bool, error) {

pipeline/frontend/yaml/constraint/constraint.go#L68
match, err := c.Match(metadata, global, env)

pipeline/frontend/yaml/constraint/constraint.go#L143
func (c *Constraint) Match(m metadata.Metadata, global bool, env map[string]string) (bool, error) {

pipeline/frontend/yaml/constraint/constraint.go#L172-174
if c.Evaluate != "" {
	maps.Copy(env, m.Environ())
	out, err := expr.Compile(c.Evaluate, expr.Env(env), expr.AsBool())
Evaluation against built-in and custom env variables

server/pipeline/create.go#L131
pipeline, pipelineItems, err := createPipelineItems(ctx, _store, pipeline, repoUser, repo, forgeYamlConfigs, nil)
envs is nil always, but in pipeline restarating (server/api/pipeline.go#L470)

server/pipeline/items.go#L31
func createPipelineItems(c context.Context, store store.Store, currentPipeline *model.Pipeline, user *model.User, repo *model.Repo,
	yamls []*forge_types.FileMeta, envs map[string]string,) (*model.Pipeline, []*pipeline.Item, error) {

server/pipeline/items.go#L61-62
for _, global := range globals {
	envs[global.Name] = global.Value
Copying service/global variables

server/pipeline/items.go#L66-67
for k, v := range currentPipeline.AdditionalVariables {
	envs[k] = v

server/pipeline/items.go#L70-82
b := pipeline.StepBuilder{
	Envs:  envs,
}
pipelineItems, err := b.Build()
Run step builder with env variables

pipeline/stepBuilder.go#L64
func (b *StepBuilder) Build() ([]*Item, error) {

pipeline/stepBuilder.go#L93
environ := b.environmentVariables(workflowMetadata, axis)
Creating env vars from step metadat and matrix

pipeline/stepBuilder.go#L101
environ[k] = v
Appending global (and additional) non-existent vars

pipeline/stepBuilder.go#L124
if match, err := parsed.When.Match(workflowMetadata, true, environ); !match && err == nil {
Checking if filtered

pipeline/stepBuilder.go#L136
ir, err := b.toInternalRepresentation(parsed, environ, workflowMetadata, workflow.ID)

pipeline/stepBuilder.go#L248-249
compiler.WithEnviron(environ),
compiler.WithEnviron(b.Envs),
Copying env vars into compiler

pipeline/stepBuilder.go#L279
).Compile(parsed)

pipeline/frontend/yaml/compiler/compiler.go#L104
func (c *Compiler) Compile(conf *yaml_types.Workflow) (*backend_types.Config, error) {

pipeline/frontend/yaml/compiler/compiler.go#L107
if match, err := conf.When.Match(c.metadata, true, c.env); !match && err == nil {
If pipeline does not match the configured filter then stop further compilation

pipeline/frontend/yaml/compiler/compiler.go#L180
if match, err := container.When.Match(c.metadata, false, c.env); !match && err == nil {
Filtering custom clone step

pipeline/frontend/yaml/compiler/compiler.go#L215
if match, err := container.When.Match(c.metadata, false, c.env); !match && err == nil {
Filtering services steps

pipeline/frontend/yaml/compiler/compiler.go#L237
if match, err := container.When.Match(c.metadata, false, c.env); !match && err == nil {
Filtering pipeline steps

pipeline/frontend/yaml/compiler/compiler.go#L253
step := c.createProcess(name, container, namePipeline)

pipeline/frontend/yaml/compiler/convert.go#L18
func (c *Compiler) createProcess(name string, container *yaml_types.Container, section string) *backend_types.Step {

pipeline/frontend/yaml/compiler/convert.go#L63
environment[k] = v
Copying compiler env vars into step env vars

pipeline/frontend/yaml/compiler/convert.go#L157
return &backend_types.Step{
	Environment:    environment,
Step creation

pipeline/backend/kubernetes/pod.go#L107
Spec: v1.PodSpec{
	Env:             mapToEnvVars(step.Environment),
Creation of Kubernets Pod with env vars from step

**Comment by zc-devs** [comment]
Tests.

```
pipeline:
  vendor:
    when:
      evaluate: 'TESTVAR == "testval"'
    image: *golang_image
    pull: true
    commands:
      - go mod vendor
```

1.
Screenshot 2023-07-01 163102
Screenshot 2023-07-01 163159
2.
Screenshot 2023-07-01 163300
Screenshot 2023-07-01 163335
3.
Screenshot 2023-07-01 163503
Screenshot 2023-07-01 163521
There is error still when the variable is not set. Let's deal with it later. But I appreciate any thoughts about possible fix.

**Comment by zc-devs** [comment]
`3` is gonna be fixed by

```go
import (
	"fmt"

	"github.com/antonmedv/expr"
)

func main() {
	env := map[string]interface{}{
		// "TESTVAR": "testval",
		// "TESTVAR": "qwe",
	}
	cEvaluate := "TESTVAR == \"testval\""

	out, err := expr.Compile(cEvaluate, expr.Env(env), expr.AllowUndefinedVariables(), expr.AsBool())
	if err != nil {
		fmt.Printf("compile err: %v\n", err)
	}
	result, err := expr.Run(out, env)
	if err != nil {
		fmt.Printf("run err: %v\n", err)
	}
	fmt.Printf("result.(bool): %v\n", result.(bool))
}
```

**Comment by zc-devs** [comment]
Test of third case:
Screenshot 2023-07-02 115131
Screenshot 2023-07-02 115227

File: docs/docs/20-usage/20-pipeline-syntax.md [hunk 1]
```diff
@@ -444,7 +444,7 @@ when:
 
 #### `evaluate`
 
-Execute a step only if the provided evaluate expression is equal to true. Each [`CI_` variable](./50-environment.md#built-in-environment-variables) can be used inside the expression.
+Execute a step only if the provided evaluate expression is equal to true. Both built-in [`CI_`](./50-environment.md#built-in-environment-variables) and custom variables can be used inside the expression.
```
Comment (6543) [inline docs/docs/20-usage/20-pipeline-syntax.md:447]: will built-in overwrite custom or other way around ... define clear

Comment (zc-devs) [reply docs/docs/20-usage/20-pipeline-syntax.md:447]: [Copied](https://github.com/woodpecker-ci/woodpecker/pull/1907/commits/1b9007af4aa5c6b17c5b2066ea1edd3e62840c68) from [Global environment variables](https://github.com/woodpecker-ci/woodpecker/blob/1b9007af4aa5c6b17c5b2066ea1edd3e62840c68/docs/docs/20-usage/50-environment.md?plain=1#L135).  [Implemented](https://github.com/woodpecker-ci/woodpecker/blob/1b9007af4aa5c6b17c5b2066ea1edd3e62840c68/pipeline/frontend/yaml/constraint/constraint.go#L174) according docs.

**Review by 6543 — APPROVED** [review]

**Pull request merged**